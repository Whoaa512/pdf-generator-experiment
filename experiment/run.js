import fs from 'fs'

import pify from 'pify'
import { VM } from 'vm2'
import { transformAsync } from '@babel/core'
import JsxPdf from 'jsx-pdf'
import PDFMake from 'pdfmake'
import getStream from 'get-stream'

import fonts from '../src/fonts'

const readFile = pify(fs.readFile)
let DEBUG = false
const DEFAULT_TIMEOUT = 5000

export const log = {
    debug(...args) {
        if (!DEBUG) {
            return
        }
        /* eslint-disable-next-line no-console */
        console.debug(...args)
    },
    error(...args) {
        /* eslint-disable-next-line no-console */
        console.error(...args)
    },
    info(...args) {
        /* eslint-disable-next-line no-console */
        console.log(...args)
    },
}

export async function run(
    filePath,
    { debug = false, timeout = DEFAULT_TIMEOUT } = {},
) {
    const jsxString = await readFile(filePath)

    if (debug === true) {
        DEBUG = true
    }

    log.debug({ jsxString }, 'read file')
    const code = `JsxPdf.renderPdf(${jsxString})`

    const compiled = await transformAsync(code, {
        plugins: [
            [
                '@babel/plugin-transform-react-jsx',
                {
                    pragma: 'JsxPdf.createElement',
                },
            ],
        ],
    })

    log.debug('~~~~~~~~')
    log.debug('transformed:', compiled.code)
    log.debug('~~~~~~~~')

    const vm = new VM({ timeout })

    vm.freeze(JsxPdf, 'JsxPdf')

    const output = vm.run(`${compiled.code}`.replace('"use strict";', ''))

    log.debug({ output }, 'evaluated result')

    const pdfMake = new PDFMake(fonts)
    const pdfStream = pdfMake.createPdfKitDocument(output)

    pdfStream.end()
    const pdfBuffer = await getStream.buffer(pdfStream)

    log.debug('pdfBuffer', pdfBuffer)
    const dataUrl = `data:application/pdf;base64,${pdfBuffer.toString(
        'base64',
    )}`

    log.debug('dataUrl', dataUrl)

    return dataUrl
}
