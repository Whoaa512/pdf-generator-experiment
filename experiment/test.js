import { run } from './run'

jest.mock('get-stream', () => ({
    buffer: jest.fn(async () => {
        return 'hello I am pdf'
    }),
}))

test('happy path', async () => {
    expect.assertions(1)

    const result = await run('./fixtures/test-jsx.txt')

    expect(result).toMatchSnapshot()
})

describe('invalid input', () => {
    test('attack-1', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-1.txt')

        await expect(result).rejects.toThrowError('Script execution timed out.')
    })

    test('attack-2', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-2.txt')

        await expect(result).rejects.toThrowError(
            "Cannot read property 'children' of undefined",
        )
    })

    test('attack-3', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-3.txt')

        await expect(result).rejects.toThrowError(
            'The <document> element can only contain <header>, <content>, and <footer> elements but found undefined',
        )
    })

    test('attack-4', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-4.txt')

        await expect(result).rejects.toThrowError(
            "Cannot read property 'elementName' of null",
        )
    })

    test('attack-5', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-5.txt')

        await expect(result).rejects.toThrowError('require is not defined')
    })

    test('attack-6', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-6.txt')

        await expect(result).rejects.toThrowError('unknown: Unexpected token')
    })

    test('attack-7', async () => {
        expect.assertions(1)

        const result = run('./fixtures/attack-7.txt')

        await expect(result).rejects.toThrowError('unknown: Unexpected token')
    })
})
