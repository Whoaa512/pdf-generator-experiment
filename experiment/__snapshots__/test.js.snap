// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy path 1`] = `"data:application/pdf;base64,hello I am pdf"`;
